DOC-4734 SSO SQL access (#15657)
cockroachdb/docs: cockroachcloud/cloud-sso-sql.md

---
title: Cluster Single Sign-on (SSO) for {{ site.data.products.db }}

> summary: Overview of Cluster Single Sign-on (SSO) for {{ site.data.products.db }}, and review of workflows for authenticating human and bot users, and for configuring required cluster settings.

Reviewer: ```suggestion
summary: Overview of Cluster Single Sign-on (SSO) for {{ site.data.products.db }}, and review of workflows for authenticating human and application users, and for configuring required cluster settings.
```

toc: true
docs_area: manage
---

> Cluster SSO allows users to access the SQL interface of a CockroachDB cluster (whether provisioned on {{ site.data.products.db }} or self-hosted) with the full security of Single Sign-On (SSO), and the convenience of being able to choose from a variety of SSO identity providers, including {{ site.data.products.db }}, Google, Microsoft, GitHub, or your own self-hosted SAML or OIDC.

Reviewer: ```suggestion
Cluster SSO allows users to access the SQL interface of a CockroachDB cluster (whether provisioned on {{ site.data.products.db }} or self-hosted) with the full security of Single Sign-On (SSO), and the convenience of being able to choose from a variety of SSO identity providers, including {{ site.data.products.db }}, Google, Azure, GitHub, or your own self-hosted OIDC.
```

{{site.data.alerts.callout_info}}
Note that this authentication method only works for human users, since only humans may have {{ site.data.products.db }} Console identities.

Software users (i.e. service accounts), can authenticate using JWT tokens from your own identity provider. See [Authenticating software users (service accounts) with Cluster SSO](#authenticate-human-users-with-the-cloud-console).
{{site.data.alerts.end}}

**Learn more:**

- [Single Sign-On (SSO) for CockroachDB Cloud organizations](cloud-org-sso.html#cloud-organization-sso)
- [Configure Cloud Organization Single Sign-On (SSO)](configure-cloud-org-sso.html)

**Prerequisites:**

- You must have a user identity on a {{ site.data.products.db }} organization. For help setting up an organization and cluster, see: [Quickstart with CockroachDB](quickstart.html).
- SSO must be enabled for your organization. For help configuring SSO for your {{ site.data.products.db }} organization, see: [Configure Cloud Organization Single Sign-On (SSO)](configure-cloud-org-sso.html)

> - SSO must be enabled for your particular {{ site.data.products.db }} user. Configure this at the [{{ site.data.products.db }} Console account settings page](https://cockroachlabs.cloud/account/profile).

Reviewer: This is relevant only when use is using Basic SSO and Cloud Org SSO is not setup for the org.

???!!! {Cameron, how does this format actually work, does the "@gmail" go in there???}

- [`ccloud`, the {{ site.data.products.db }} CLI](ccloud-get-started.html) must be installed on your local workstation.

1. First authenticate to your {{ site.data.products.db }} organization. This command will cause your workstation's default browser to open to a {{ site.data.products.db }} authentication portal. Authenticate here as you normally do to the {{ site.data.products.db }} Console. The `ccloud` utility will receive an authentication token from the browser, allowing you to authenticate from the command line.

	{% include_cached copy-clipboard.html %}
	~~~shell
	ccloud auth login
	~~~

1. You may then use the `ccloud` utility to authenticate to your {{ site.data.products.db }} cluster, allowing you to access the SQL interface.

	{% include_cached copy-clipboard.html %}
	~~~shell
	ccloud cluster sql --sso {your cluster name}
	~~~

> ## Authenticate software users (service accounts) with external IdPs

Reviewer: ```suggestion
## Authenticate application users (service accounts) with external IdPs
```

> Currently, {{ site.data.products.db }} can only serve as a token issuer for human users. Authenticating service accounts, i.e. user identities to be controlled by software applications or scripts, rather by humans, is considerably more complicated, as it requires the user, or more realistically, an IdP admin, to provision the appropriate JWT token.

Reviewer: ```suggestion
Currently, {{ site.data.products.db }} can only serve as a token issuer for human users as it requires an interactive flow. Authenticating service accounts, i.e. user identities to be controlled by software applications or scripts, rather than by humans, can not leverage an interactive flow and it needs a different headless mechanism.
```

> {{ site.data.products.db }} SSO supports the use of external IdPs such as Google, Microsoft, GitHub, or customer deployed OIDC or SAML solutions, such as Okta. All of these options support multifactor authentication (MFA).

Reviewer: Are we referring to Org SSO here or the cluster SSO with service accounts? It looks like for former within the section for latter. MFA doesn't apply to cluster SSO with service accounts and the list of validated providers is different - Google Cloud Platform Service Accounts, Azure Managed Identities, AWS EKS Service Accounts and Okta identities.

> {{ site.data.products.db }} SSO supports the use of external IdPs such as Google, Microsoft, GitHub, or customer deployed OIDC or SAML solutions, such as Okta. All of these options support multifactor authentication (MFA).

Author: ```suggestion
{{ site.data.products.db }} console SSO supports the use of external IdPs such as Google, Microsoft, GitHub, or customer deployed OIDC or SAML solutions, such as Okta. All of these options support multifactor authentication (MFA). This applies to authenticating to {{ site.data.products.db }} organizations, and to authenticating to clusters using the console.
```

Author: @abhinavg6 ^^ clearer?

Reviewer: Yeah that's almost there Michael - should probably make it "This applies authentication to {{ site.data.products.db }} organizations, and to authenticating human users to clusters through the console".

But I wonder if it's in the right place? This section is for application / service account authentication with external IdPs, where it's not relevant.

**Learn more:**

> This [Cockroach Labs blog post](https://www.cockroachlabs.com/blog/) covers and provides further resources for a variety of auth token-issuing use cases, including using Okta and Google Cloud Platform to issue tokens.

Reviewer: If we are looking to refer to the blog post that I had shared earlier, it won't be out until end of next week. So we may need to remove it for a bit as we need to publish the doc soon. We could add it back later.

**Prerequisites:**

- You must have the ability to create identities and issue tokens with a SAML or OIDC identity provider.
	{{site.data.alerts.callout_success}}
	The `issuer` and `audience` fields in the token must match values [configured in your cluster settings](#configure-your-cluster-to-accept-your-external-identity-provider).
	{{site.data.alerts.end}}

> - You must have a user identity on a {{ site.data.products.db }} organization. For help setting up an organization and cluster, see: [Quickstart with CockroachDB](quickstart.html).

Reviewer: This is not needed. Right @kpatron-cockroachlabs ?

> - You must have access to a member of the [`admin` role](../{{site.versions["stable"]}}/security-reference/authorization.html#admin-role) in order to update cluster settings, which is required to add an external token issuer/IdP.

Reviewer: Actually it can be done by just having the new MODIFYCLUSTERSETTING permission as well - https://www.cockroachlabs.com/docs/stable/security-reference/authorization.html#supported-privileges

> - You must have access to a member of the [`admin` role](../{{site.versions["stable"]}}/security-reference/authorization.html#admin-role) in order to update cluster settings, which is required to add an external token issuer/IdP.

Author: ```suggestion
- You must have access to a SQL user with either the [`admin` role](../{{site.versions["stable"]}}/security-reference/authorization.html#admin-role) or the [`MODIFYCLUSTERSETTING`](permissions), in order to update cluster settings. This is required to add an external token issuer/IdP.
```

- A SQL user specifically corresponding to the service account must be pre-provisioned on the cluster (or you must have access to a SQL role allowing you to create such a user).

### Configure your cluster to accept your external identity provider

In order to authenticate a service account to a {{ site.data.products.db }} cluster using a JWT issuer, you must update several cluster settings in the `server.jwt_authentication` namespace:

1. `server.jwt_authentication.jwks`

	Add your IdP's public signing key to your cluster's list of accepted signing JSON web keys (JWKS), under the `jwks` setting. This is a [JWK](https://www.rfc-editor.org/rfc/rfc7517) formatted single key or key set, containing the [public keys](../{{site.versions["stable"]}}/security-reference/transport-layer-security.html#key-pairs) for SSO token issuers/IdPs that will be accepted by your cluster.

> 	By default, your cluster's configuration will contain the {{ site.data.products.db }}'s own public key, allowing {{ site.data.products.db }} to serve as an IdP. When modifying this cluster setting, you must include the {{ site.data.products.db }} public key in the key set. Failing to do so can prevent maintenance access by essential {{ site.data.products.db }} managed service accounts, leading to unintended consequences.

Reviewer: ```suggestion
	By default, your cluster's configuration will contain the {{ site.data.products.db }}'s own public key, allowing {{ site.data.products.db }} to serve as an IdP. When modifying this cluster setting, you must include the {{ site.data.products.db }} public key in the key set. Failing to do so can prevent maintenance access by the {{ site.data.products.db }} managed service, leading to unintended consequences. It can also break the cluster SSO for human users to this cluster.
```

	!!!{ @cameron Fact check on this? Seems right}

1. `server.jwt_authentication.issuers`

	Add your IdP's formal `issuer` name (this must match the `issuer` field in the JWT itself) to your cluster's list of accepted token issuers. This field takes a comma-separated list of formal names of accepted JWT issuers. This list must include a given IdP, or the cluster will reject JWTs issued by it.

> 	The default value is {{ site.data.products.db }}. When modifying this cluster setting, you must include {{ site.data.products.db }} in the new value. Failing to do so can prevent maintenance access by essential {{ site.data.products.db }} managed service accounts, leading to unintended consequences.

Reviewer: ```suggestion
	The default value is {{ site.data.products.db }}. When modifying this cluster setting, you must include {{ site.data.products.db }} in the new value. Failing to do so can prevent maintenance access by the {{ site.data.products.db }} managed service, leading to unintended consequences. It can also break the cluster SSO for human users to this cluster.
```

	!!!{ @cameron Fact check on this? Seems right}

1. `server.jwt_authentication.audience`
	
	The name of your cluster as specified by the IdP, or a comma-separated list of such names. One of the audience names configured here must match the `audience` field with which your IdP will generate JWT formatted auth tokens.

1. `server.jwt_authentication.enabled`

	Required value: `true`

	Check to confirm that JWT authentication is enabled on your cluster. It is enabled by default in {{ site.data.products.db }} clusters.

### Authenticate to your cluster with your JWT token

> To provision SQL cluster access for service accounts, you must provision OIDC or SAML tokens. There are many ways to do this, which are beyond the scope of this tutorial.

Reviewer: ```suggestion
To provision SQL cluster access for application users or service accounts, you must provision JWT tokens. There are many ways to do this, which are beyond the scope of this tutorial.
```

> For example, your Google Cloud Platform organization can serve as IdP by issuing OIDC auth tokens, as described here in the [GCP docs on issuing tokens to service accounts](https://cloud.google.com/iam/docs/create-short-lived-credentials-direct#sa-credentials-oidc). This [blog post](https://morgans-blog.deno.dev/sso-crdb-gcp) discusses using GCP-issued OIDC tokens to authenticate to CockroachDB.

Reviewer: ```suggestion
For example, your Google Cloud Platform organization can serve as IdP by issuing OIDC auth tokens for GCP service accounts, as described here in the [GCP docs on issuing tokens to service accounts](https://cloud.google.com/iam/docs/create-short-lived-credentials-direct#sa-credentials-oidc). This [blog post](https://morgans-blog.deno.dev/sso-crdb-gcp) discusses using GCP-issued OIDC tokens to authenticate to CockroachDB.
```

> ---

Author: fixed

title: Cluster Single Sign-On (SSO) for Cockroach Cloud
summary: Overview of Cluster Single Sign-On (SSO) for {{ site.data.products.db }}, and review of workflows for authenticating human and bot users, and for configuring required cluster settings.
toc: true
docs_area: manage
---

> ## What is Cluster Single Sign-On (SSO)?

Reviewer 2: Same for the self-hosted page!

> ## What is Cluster Single Sign-On (SSO)?

Author: done

Cluster SSO allows users to access the SQL interface of a CockroachDB cluster (whether provisioned on {{ site.data.products.db }} or self-hosted) with the full security of Single Sign-On (SSO), and the convenience of being able to choose from a variety of SSO identity providers, including Cockroach Cloud, Google, Microsoft, GitHub, or your own self-hosted SAML or OIDC.

This page describes:

- [Authenticating human users](#authenticate-human-users-with-the-cloud-console) with the {{ site.data.products.db }} console acting as identity provider (IdP) or token issuer
- [Authenticating service accounts](#authenticate-software-users-service-accounts-with-external-idps) with an external token issuer
- [Configuring the cluster settings](#configure-your-cluster-to-accept-your-external-identity-provider) required to authenticate service accounts

> Note that this regards SQL access to a specific CockroachDB Cluster, not access to a Cockroach Cloud organization. For the latter, see [Single Sign-On (SSO) for CockroachDB Cloud organizations](cloud-org-sso.html).

Author: yup all the clouds should be liquidized, done.

> Note that this regards SQL access to a specific CockroachDB Cluster, not access to a Cockroach Cloud organization. For the latter, see [Single Sign-On (SSO) for CockroachDB Cloud organizations](cloud-org-sso.html).

Author: has been updated

You might also be looking for [Cluster Single Sign-On (SSO) for Self-hosted CockroachDB](../v22.2/sso-sql.html).

> ## Authenticate human users with the cloud console

Author: done

{{ site.data.products.db }} can issue authentication tokens for human users, which can be used for SSO using the `ccloud` CLI. {{ site.data.products.dedicated }} and {{ site.data.products.serverless }} clusters are, by default, configured to allow Cluster SSO authentication with {{ site.data.products.db }} serving as SSO provider. No additional set-up is required.

{{site.data.alerts.callout_info}}
Note that this authentication method only works for *human users*, since only humans may have {{ site.data.products.db }} console identities.

Software users (i.e. service accounts), can authenticate using JWT tokens from your own identity provider. See [Authenticating software users (service accounts) with Cluster SSO](#authenticate-human-users-with-the-cloud-console).
{{site.data.alerts.end}}

> **Learn more:**

Reviewer 2: Not totally committed to this comment: but what do you think about switching "Learn more" and "Prerequisites" to H3 headers?

Author: I think it clutters the nav too much, these shouldn't show up in the nav IMO

???!!! {Cameron, how does this format actually work, does the "@gmail" go in there???}

- [`ccloud`, the Cockroach Cloud CLI](ccloud-get-started.html) must be installed on your local workstation.

1. First authenticate to your {{ site.data.products.db }} organization. This command will cause your workstation's default browser to open to a Cockroach Cloud authentication portal. Authenticate here as you normally do to the Cockroach Cloud console. The `ccloud` utility will receive an authentication token from the browser, allowing you to authenticate from the command line.

	{% include_cached copy-clipboard.html %}
	~~~shell
	ccloud auth login
	~~~

1. You may then use the `ccloud` utility to authenticate to your {{ site.data.products.db }} cluster, allowing you to access the SQL interface.

	{% include_cached copy-clipboard.html %}
	~~~shell

> 	ccloud cluster sql --sso { your cluster name}

Author: good catch!

	~~~

## Authenticate software users (service accounts) with external IdPs

Currently, Cockroach Cloud can only serve as an token issuer for human users. Authenticating service accounts, i.e. user identities to be controlled by software applications or scripts, rather by humans, is considerably more complicated, as it requires the user, or more realistically, an IdP admin, to provision the appropriate JWT token.

Cockroach Cloud SSO supports the use of external IdPs such as Google, Microsoft, Github, or customer deployed OIDC or SAML solutions, such as Okta. All of these options support multifactor authentication (MFA).

> **Learn more:**

Reviewer 2: The format of these sections with Learn more and Prerequisites is super clear!

Author: thanks!

### Configure your cluster to accept your external identity provider

In order to authenticate a service account to a {{ site.data.products.db }} cluster using a JWT issuer, you must update several cluster settings in the `server.jwt_authentication` namespace:

1. `server.jwt_authentication.jwks`

	Add your IdP's public signing key to your cluster's list of accepted signing JSON web keys (JWKS), under the `jwks` setting.

	This is a [JWK](https://www.rfc-editor.org/rfc/rfc7517) formatted single key or key set, containing the [public keys](../{{site.versions["stable"]}}/security-reference/transport-layer-security.html#key-pairs) for SSO token issuers/IdPs that will be accepted by your cluster.

	By default, your cluster's configuration will contain the {{ site.data.products.db }}'s own public key, allowing Cockroach Cloud to serve as an IdP.

	When modifying this cluster setting, you must include the Cockroach Cloud public key in the key set. Failing to do so can prevent maintenance access by essential Cockroach Cloud managed service accounts, leading to unintended consequences.

	!!!{ @cameron Fact check on this? Seems right}

> 1. `server.jwt_authentication.issuers`

Author: improved a little

> 	Add your IdP's formal `issuer` name (this must match the `issuer` field in the JWT itself) to your cluster's list of accepted token issuers:

Author: fixed

> 	A comma separated list of formal names of accepted JWT issuers. This list must include a given IdP, or the cluster will reject JWTs issued by it.

Reviewer 2: Nit: "comma separated list" --> "comma-separated list"

Author: fixed

> 	By default will be the CC issuer but they can set it to any string or any list of strings formatted as a json array. Needs to match the issuer of JWTs

Author: fixed

1. `server.jwt_authentication.enabled`

	Required value: `true`

	Check to confirm that JWT authentication is enabled on your cluster. It is enabled by default in {{ site.data.products.db }} clusters.

### Authenticate to your cluster with your JWT token

To provision SQL cluster access for service accounts, you must provision OIDC or SAML tokens. There are many ways to do this, which are beyond the scope of this tutorial.

For example, your Google Cloud Platform organization can serve as IdP by issuing OIDC auth tokens, as described here in the [GCP docs on issuing tokens to service accounts](https://cloud.google.com/iam/docs/create-short-lived-credentials-direct#sa-credentials-oidc). This [blog post](https://morgans-blog.deno.dev/sso-crdb-gcp) discussing using GCP-issued OIDC tokens to authenticate to CockroachDB.

Once you have a valid JWT auth token (with `issuer` and `audience` matching the values [configured in your cluster settings](#configure-your-cluster-to-accept-your-external-identity-provider)) from your IdP, you may use it to connect to your cluster's SQL interface.

> {{site.data.alerts.success}}

Reviewer 3: ```suggestion
{{site.data.alerts.callout_success}}
```

This example uses [`cockroach sql`](../{{site.versions["stable"]}}/cockroach-sql.html), but you can use any SQL client that supports sufficiently long passwords.
{{site.data.alerts.end}}

{% include_cached copy-clipboard.html %}
~~~shell
cockroach sql --url "postgresql://{SQL_USERNAME}:{JWT_TOKEN}@{CLUSTER_HOST}:26257?options=--crdb:jwt_auth_enabled=true" --certs-dir={CLUSTER_CERT_DIR}
~~~

~~~txt
Welcome to the cockroach SQL interface...
~~~

> ## Support for tokens with SQL-incompatible subject names

Reviewer 2: Did you intend for this heading to be a H2? It seems to not quite fit with the page structure.

Author: It's less of a full section, but it doesn't nest under the others, so yeah, it gets its own h2 as far as I can see.

---
title: Cluster Single Sign-on (SSO) for CockroachDB Cloud
summary: Overview of Cluster Single Sign-on (SSO) for {{ site.data.products.db }}, review of authenticating users, configuring required cluster settings.
toc: true
docs_area: manage
---

Cluster SSO allows users to access the SQL interface of a CockroachDB cluster (whether provisioned on {{ site.data.products.db }} or self-hosted) with the full security of Single Sign-On (SSO), and the convenience of being able to choose from a variety of SSO identity providers, including {{ site.data.products.db }}, Google, Azure, GitHub, or your own self-hosted OIDC.

This page describes:

- [Authenticating human users](#authenticate-human-users-with-the-cloud-console) with the {{ site.data.products.db }} Console acting as identity provider (IdP) or token issuer
- [Authenticating service accounts](#authenticate-application-users-service-accounts) with an external token issuer

> - [Configuring the cluster settings](#configure-your-cluster-to-accept-your-external-identity-provider) required to authenticate service accounts

Reviewer 4: Thinking we could probably have these sections on a different page since this page is about Cloud.

Note that this regards SQL access to a specific CockroachDB Cluster, not access to a {{ site.data.products.db }} organization. For the latter, see [Single Sign-On (SSO) for {{ site.data.products.db }} organizations](cloud-org-sso.html).

You might also be looking for [Cluster Single Sign-On (SSO) for Self-hosted CockroachDB](../v22.2/sso-sql.html).

> ## Authenticate human users with the cloud console

Reviewer 4: Could we change this section name to something like "Getting started" or just removing it altogether? At the top, we could link to an external tokens page for authenticating service accounts.

- To authenticate to a specific cluster using SSO, a {{ site.data.products.db }} user must have a corresponding SQL user already [created](../{{site.versions["stable"]}}/create-user.html#create-a-user) on that cluster. {{ site.data.products.db }} generates a SSO SQL username for each console, corresponding to the user's email by the convention `sso_{email_name}`, where `email_name` is everything up to the `@` in an email address, for example the SQL user `sso_docs` would result from `docs@cockroachlabs.com`. `ccloud` will prompt you to make this user if it does not already exist, in which case an admin must create it manually. 
- [`ccloud`, the {{ site.data.products.db }} CLI](ccloud-get-started.html) must be installed on your local workstation.

1. First authenticate to your {{ site.data.products.db }} organization. This command will cause your workstation's default browser to open to a {{ site.data.products.db }} authentication portal. Authenticate here as you normally do to the {{ site.data.products.db }} Console. The `ccloud` utility will receive an authentication token from the browser, allowing you to authenticate from the command line.

> 

Reviewer 4: For the steps here, I would try to make this section align as closely as possible to [this internal guide that we have](https://cockroachlabs.atlassian.net/wiki/spaces/CDX/pages/2821062718/Cloud+Cluster+SSO+Guide).

---
title: Cluster Single Sign-on (SSO) using CockroachDB Cloud Console

> summary: Overview of Cluster Single Sign-on (SSO) for {{ site.data.products.db }}, review of authenticating users, configuring required cluster settings.

Author: There are a lot of "Overview..." seems like a common way to begin these summaries, which are not publicly visible anyway, correct?

toc: true
docs_area: manage
---

> Cluster SSO allows users to access the SQL interface of a CockroachDB cluster (whether provisioned on {{ site.data.products.db }} or self-hosted) with the full security of Single Sign-On (SSO), and the convenience of being able to choose from a variety of SSO identity providers, including {{ site.data.products.db }}, Google, Azure, GitHub, or your own self-hosted OIDC.

Author: no, I meant OIDC.

> This page describes the procedure for accessing a {{ site.data.products.db }} cluster using the {{ site.data.products.db }} console as identity provider.

Author: I prefer my wording for the second part. The console isn't a storage unit lol

> This page describes the procedure for accessing a {{ site.data.products.db }} cluster using the {{ site.data.products.db }} console as identity provider.

Reviewer 5: OK

It is also possible to authenticate using JWT tokens from your external customer-managed IdP. See: [Cluster Single Sign-on (SSO) using JSON web tokens (JWT)](../{{site.versions["stable"]}}/sso-sql.html)

{{site.data.alerts.callout_info}}
Note that this authentication method only works for human users, since only humans may have {{ site.data.products.db }} Console identities.

Application users (i.e. service accounts), can authenticate using JWT tokens from your own identity provider. See [Cluster Single Sign-on (SSO) using JSON web tokens (JWT)](../{{site.versions["stable"]}}/sso-sql.html).

> Note that this regards SQL access to a specific CockroachDB Cluster, not access to a {{ site.data.products.db }} organization. For the latter, see [Single Sign-On (SSO) for {{ site.data.products.db }} organizations](cloud-org-sso.html).

Author: SSO isn't an access management tool here, it's an authentication method, so your suggested wording doesn't seem right.

> Note that this regards SQL access to a specific CockroachDB Cluster, not access to a {{ site.data.products.db }} organization. For the latter, see [Single Sign-On (SSO) for {{ site.data.products.db }} organizations](cloud-org-sso.html).

Reviewer 5: I reworked your rework a bit. PTAL.

{{site.data.alerts.end}}

## Prerequisites

- You must have a user identity on a {{ site.data.products.db }} organization, with access to a cluster or the ability to create a cluster. For help setting up an organization and cluster, see: [Quickstart with CockroachDB](quickstart.html).

> - To authenticate to a specific cluster using SSO, a {{ site.data.products.db }} user must have a corresponding SQL user already [created](../{{site.versions["stable"]}}/create-user.html#create-a-user) on that cluster. {{ site.data.products.db }} generates a SSO SQL username for each console, corresponding to the user's email by the convention `sso_{email_name}`, where `email_name` is everything up to the `@` in an email address, for example the SQL user `sso_docs` would result from `docs@cockroachlabs.com`. `ccloud` will prompt you to make this user if it does not already exist, in which case an admin must create it manually. 

Author: Well, it depends on whether the user of the docs is trying to do the provisioning or just trying to achieve access. These docs may be for users who are just trying to do the accessing, not for admins who are provisioning access. For the former, the stuff must be pre-provisioned.

> - To authenticate to a specific cluster using SSO, a {{ site.data.products.db }} user must have a corresponding SQL user already [created](../{{site.versions["stable"]}}/create-user.html#create-a-user) on that cluster. {{ site.data.products.db }} generates a SSO SQL username for each console, corresponding to the user's email by the convention `sso_{email_name}`, where `email_name` is everything up to the `@` in an email address, for example the SQL user `sso_docs` would result from `docs@cockroachlabs.com`. `ccloud` will prompt you to make this user if it does not already exist, in which case an admin must create it manually. 

Author: It's a confusing mess of a doc overall and could be much improved by being broken out into separate tutorials, etc., if there were more time.

> - [`ccloud`, the {{ site.data.products.db }} CLI](ccloud-get-started.html) must be installed on your local workstation.

Author: Pretty sure can't use this method on a VM, since a VM doesn't normally have a browser that you can log into, does it?

> - [`ccloud`, the {{ site.data.products.db }} CLI](ccloud-get-started.html) must be installed on your local workstation.

Reviewer 5: I don't know why it couldn't. Aside from the fact that a VM can have and serve a UI (I am currently doing some testing on a Windows 10 VM), command-line browsers (`curl`, `wget`, `lynx`....) also support authentication.

> - [`ccloud`, the {{ site.data.products.db }} CLI](ccloud-get-started.html) must be installed on your local workstation.

Author: nah your right, that's what that redirect flag is for as well.

## Signing in with Cluster SSO

> 1. First authenticate to your {{ site.data.products.db }} organization. This command will cause your workstation's default browser to open to a {{ site.data.products.db }} authentication portal. Authenticate here as you normally do to the {{ site.data.products.db }} Console. The `ccloud` utility will receive an authentication token from the browser, allowing you to authenticate from the command line.

Reviewer 5: ```suggestion
1. Authenticate the `ccloud` command to your {{ site.data.products.db }} organization. When you run the following command, your workstation's default browser opens the {{ site.data.products.db }}  login page for your organization.

	{% include_cached copy-clipboard.html %}
	~~~shell
	ccloud auth login --org {your organization label}
	# when running ccloud on a remote host, add the following:
	# --no-redirect
	~~~
 
1. Sign in using your {{ site.data.products.db }} identity. The `ccloud` utility receives an authentication token from the browser and uses it to authenticate itself to your {{ site.data.products.db }} organization.
```

Have you shown them how to get their organization label?

> 	{% include_cached copy-clipboard.html %}

Reviewer 5: If you take my edit above, then you can remove lines 33-38 here.

	~~~shell
	ccloud auth login --org {your org label}
	# when running ccloud on a remote host, add the following:
	# --no-redirect
	~~~

> 1. You may then use the `ccloud` utility to authenticate to your {{ site.data.products.db }} cluster, allowing you to access the SQL interface. Your browser will open again as `ccloud` requests a fresh token, although will not need to log in again if you are already logged in.

Reviewer 5: This one might benefit from rewriting to be parallel with the previous suggestions if you take them. I was a bit confused about what to expect about the browser opening again but not needing to log in again -- what will they see?

Author: I don't know, I haven't been able to use the feature

	{% include_cached copy-clipboard.html %}
	~~~shell
	ccloud cluster sql --sso {your cluster name}
	# when running ccloud on a remote host, add the following:
	# --no-redirect
	~~~

## Troubleshooting

> If you get an error that the SQL user is missing, you may need to create it manually. The SQL user's username **must** match the {{ site.data.products.db }} identity's email address, according to the following convention:

Reviewer 5: Maybe give what the error actually is, to help Googlers find the info. If you take my suggestion above about adding a section for creating or finding the SQL SSO user, then you could link from here to that section.

Author: That would be great for a follow up ticket when I can actually use the functionality.
